### ISSUE TITLE
PL does not support twig extends

### ISSUE THREAD

**Issue by gael-boyenval** [pattern-lab/patternlab-node#554]
I am using Pattern Lab Node `v2.6.1` on `Mac`, with Node `v6.7`, using the `Gulp` Edition.

##### Expected Behavior
when extending a template, it extend the template… 
it is referenced here too : https://github.com/pattern-lab/patternlab-node/issues/285

##### Actual Behavior
when extending, pl make it like including

##### Steps to Reproduce
x.twig
```
foo
{% block content %}
    bar
{% endblock %}
```
=> output `foobar`

y.twig
```
{% extends "x.twig" %}
foo
{% block content %}
    baz
{% endblock %}
```
=> output `foobar` to
=> should output `foobaz`

z.twig
```
{% extends "x.twig" %}
foo
{% block content %}
    {{ parent() }}
    baz
{% endblock %}
```
=> output `foobarbar`
=> should output `foobarbaz`

**Comment by max-dombrowski** [comment]
I have the same issue. Any progress on this?

**Comment by bmuenzenmeyer** [comment]
Unfortunately, I doubt it/

Our initial contributor to the Twig engine has not been seen in a while

Cross - linking to https://github.com/pattern-lab/patternlab-node/issues/285

**Comment by KatieMFritz** [comment]
Has anyone found a workaround for this? Maybe with [Nunjucks](https://mozilla.github.io/nunjucks/)? I wish I could fix it but my JS and PHP chops aren't there yet. 😕

**Comment by gael-boyenval** [comment]
@KatieMFritz @bmuenzenmeyer. 

I looked into it quickly. 
It could be possible to use the twig engine but the current problem is that it is not possible to use inline templates when including/extending twig templates in one another.

So we have to use templates file path in order to render them. 

@bmuenzenmeyer : finally it only require for the object param passed to the renderPattern method to give us the root template path (the user's `_pattern` directory). It is currently not the case. 
If you can do that I could update the twig pattern-engine in order to make it work.

**Comment by KatieMFritz** [comment]
@Gael-boyenval, thanks for the quick response!

I played around with Nunjucks for a while last night, but I keep getting an error with it so I'm stuck. I'll report back if I get it working.

**Comment by bmuenzenmeyer** [comment]
@gael-boyenval thanks for the sluething and the direction - I can see what I can do :)

**Comment by gael-boyenval** [comment]
@KatieMFritz : nunjucks is another template engine. If you want it to work you'll have to develop a pattern-engine, a pattern-lab specific component on top of nunjucks.

**Comment by bmuenzenmeyer** [comment]
I spent too long trying to get this to work tonight. No combination of the below seemed to work:

```js
  renderPattern: function renderPattern(pattern, data) {
    var result = twig({
      allowInlineIncludes: true,
      path: path.relative(__dirname, path.resolve(process.cwd(), 'source', '_patterns', pattern.relPath)
      //data: pattern.template
    }).render(data);
    return result;
  },
```

cannot call render() on undefined.
barring the hard-codes, I may try to swap to node-twig next

**Comment by sghoweri** [comment]
@bmuenzenmeyer if you wanted to assign this one to me I can take a crack at this this week

**Comment by bmuenzenmeyer** [comment]
> if you wanted to assign this one to me I can take a crack at this this week

gopher it!

**Comment by sghoweri** [comment]
@bmuenzenmeyer taking another look at this now!

**Comment by danwhite85** [comment]
I'm working on a Nunjucks engine that's nearly complete. I resolved this issue by...

1. Setting expandPartials to false. FYI, this disables inline templates.

2. Since inline templates are disabled the full pattern path must be supplied for nunjucks to work. With inspiration from the underscore engine, I created a partialRegistry array that keeps track of pattern.patternPartial and pattern.relPath in the registerPartial function. 

3. In renderPattern I iterate the partialRegistry and replace instances of patternPartial with the relPath in pattern.extendedTemplate. 

May not be the best solution, but something that's working at least for now.
